discoverd/health/register_test.go

```diff
@@ -0,0 +1,298 @@
+			},
+			closeFn: func() error {
+				unregisterChan <- called{}
+				return nil
+			},
+			setMetaFn: func(meta map[string]string) error {
+				success := make(chan bool)
+				metaChan <- called{
+					args: map[string]interface{}{
+						"meta": meta,
+					},
+					returnVal: success,
+				}
+				if !<-success {
+					return errors.New("SetMeta failed")
+				}
+				return nil
+			},
+		}
+
+		registrarChan := make(chan called)
+		registrar := RegistrarFunc(func(service string, inst *discoverd.Instance) (discoverd.Heartbeater, error) {
+			success := make(chan bool)
+			registrarChan <- called{
+				args: map[string]interface{}{
+					"service": service,
+					"inst":    inst,
+				},
+				returnVal: success,
+			}
+			defer func() { registrarChan <- called{} }()
+			if <-success {
+				return heartbeater, nil
+			}
+			return nil, errors.New("registrar failure")
+		})
+
+		monitorChan := make(chan bool)
+		monitor := func(c Check, ch chan MonitorEvent) stream.Stream {
+			stream := stream.New()
+			go func() {
+				defer close(ch)
+			outer:
+				for {
+					select {
+					case up, ok := <-monitorChan:
+						if !ok {
+							break outer
+						}
+						if up {
+							ch <- MonitorEvent{
+								Check:  check,
+								Status: MonitorStatusUp,
+							}
+						} else {
+							ch <- MonitorEvent{
+								Check:  check,
+								Status: MonitorStatusDown,
+							}
+						}
+					case <-stream.StopCh:
+						break outer
+					}
+				}
+			}()
+			return stream
+		}
+
+		reg := Registration{
+			Service: "test",
+			Instance: &discoverd.Instance{
+				Meta: make(map[string]string),
+			},
+			Registrar: registrar,
+			Monitor:   monitor,
+			Check:     check,
+			Events:    make(chan MonitorEvent),
+		}
+		hb := reg.Register()
+		defer hb.Close()
+
+		errCh := make(chan bool)
+		errCheck := func(ch chan called, stop chan bool) {
+			go func() {
+				select {
+				case _, ok := <-ch:
+					if !ok {
+						return
+					}
+					errCh <- true
+				case <-stop:
+					errCh <- false
+				}
+			}()
+		}
+
+		var stop chan bool
+		currentMeta := make(map[string]string)
+		for _, step := range steps {
+			stop = make(chan bool)
+			var wait int
+
+			if step.event {
+				monitorChan <- step.up
+			}
+			if step.register {
+				call := <-registrarChan
+				c.Assert(call.args["inst"].(*discoverd.Instance).Meta, DeepEquals, currentMeta)
+				call.returnVal <- step.success
+				<-registrarChan
+			} else {
+				wait++
+				errCheck(registrarChan, stop)
+			}
+			if step.unregister {
+				// before unregistering, Addr should not be nil
+				c.Assert(hb.Addr(), Not(Equals), "")
+				select {
+				case <-unregisterChan:
+				case <-time.After(3 * time.Second):
+					c.Error("Timed out waiting for unregistration")
+				}
+				// Addr should be nil now
+				c.Assert(hb.Addr(), Equals, "")
+			} else {
+				wait++
+				errCheck(unregisterChan, stop)
+			}
+			if step.setMeta {
+				go func() {
+					call := <-metaChan
+					call.returnVal <- step.success
```

Comment:
Oh, good point... The logic for this is getting fairly complicated though, I think it should be obvious from the implementation that SetMeta will work correctly.